### ISSUE TITLE
Some questions about the mLSTM code.

### ISSUE THREAD

**Issue by StarPlatinumDa** [NX-AI/vision-lstm#17]
First of all, thank you so much for your academic contributions!
The first question is: 
The parallel version code, which is the code in Vision lstm, has corresponding formulas in the appendix of the paper. However, it does not include the m state and n state used for stabilization and normalization in the main text of the paper.  Is it because stabilization and normalization have been incorporated but not explicitly stated? I raised this question because there are m state and n state in the main text, but not here
The second one is:
Is the computational complexity of the parallel version higher than that of the recurrent version? If I want to reduce the computational complexity on long sequences, is it better to use the recurrent version of the code?

I would appreciate it if you would answer my questions.

**Comment by BenediktAlkin** [comment]
Hi, in the [parallel forward code](https://github.com/NX-AI/vision-lstm/blob/main/vision_lstm/vision_lstm.py#L49) there is the [log_D_matrix and log_D_matrix_stabilized](https://github.com/NX-AI/vision-lstm/blob/main/vision_lstm/vision_lstm.py#L108) which are the parallel versions of the stabilized gates. Then later, the parallel [normalizer state is calculated](https://github.com/NX-AI/vision-lstm/blob/main/vision_lstm/vision_lstm.py#L124C25-L124C64) and the [C_matrix is normalized](https://github.com/NX-AI/vision-lstm/blob/main/vision_lstm/vision_lstm.py#L126).

L: sequence length
D: hidden dimension
There is a tradeoff between the recurrent formulation (complexity O(LD^2)) and the parallel formulation (complexity O(L^2D). So ideally one would use a chunkwise formulation, that combines these two. However, in plain pytorch, the parallel training is the fastest one in actual runtime, because it can utilize the GPU the best. With a CUDA kernel, things might look different.

**Comment by StarPlatinumDa** [comment]
> Hi, in the [parallel forward code](https://github.com/NX-AI/vision-lstm/blob/main/vision_lstm/vision_lstm.py#L49) there is the [log_D_matrix and log_D_matrix_stabilized](https://github.com/NX-AI/vision-lstm/blob/main/vision_lstm/vision_lstm.py#L108) which are the parallel versions of the stabilized gates. Then later, the parallel [normalizer state is calculated](https://github.com/NX-AI/vision-lstm/blob/main/vision_lstm/vision_lstm.py#L124C25-L124C64) and the [C_matrix is normalized](https://github.com/NX-AI/vision-lstm/blob/main/vision_lstm/vision_lstm.py#L126).
> 
> L: sequence length D: hidden dimension There is a tradeoff between the recurrent formulation (complexity O(LD^2)) and the parallel formulation (complexity O(L^2D). So ideally one would use a chunkwise formulation, that combines these two. However, in plain pytorch, the parallel training is the fastest one in actual runtime, because it can utilize the GPU the best. With a CUDA kernel, things might look different.

Thanks, this perfectly solves my problem.
